report/3.Theory/OIL.tex

```diff
@@ -0,0 +1,137 @@
+% irisa223: http://www.irisa.fr/alf/downloads/puaut/TPNXT/images/os223.pdf
+%rms: http://lejos-osek.sourceforge.net/rms.htm
+
+\section{OSEK Implementation Language}
+An NXTOSEK program consists of two parts: An OIL (OSEK Implementation Language) file which describes the architecture of the scheduler and tasks, and a series of C/C++ source codes files defining the behavior of the system.
+In this section the content of the OIL file will be elaborated upon in order to clarify how it relates to defining a scheduler and tasks for later use.
+Below an example of a hello world application for NXTOSEK can be seen:
+\begin{lstlisting}[caption=OIL]
+#include "implementation.oil"
+CPU ATMEL_AT91SAM7S256 {
+  OS LEJOS_OSEK {
+    STATUS = EXTENDED;
+    STARTUPHOOKS = FALSE;
+    ERRORHOOK = FALSE;
+    SHUTDOWNHOOK = FALSE;
+    PRETASKHOOK = FALSE;
+    POSTTASKHOOK = FALSE;
+    USEGETSERVICEID = FALSE;
+    USEPARAMETERACCESS = FALSE;
+    USERESSCHEDULER = FALSE;
+  };
+  
+  APPMODE appmode1{};
+
+  TASK task1 {
+    AUTOSTART = TRUE {
+      APPMODE = appmode1;    
+    };
+    PRIORITY = 1;
+    ACTIVATION = 1;
+    SCHEDULE = FULL;
+    STACKSIZE = 512;
+  }
+}
+\end{lstlisting}
+
+\begin{lstlisting}[caption=C source]
+#include "kernel.h"
+#include "ecrobot_interface.h"
+
+void user_1ms_isr_type2(void){}
+
+TASK(task1){
+  while(1){
+    ecrobot_status_monitor("Hello, World!");
+    systick_wait_ms(500);
+  }
+}
+\end{lstlisting}
+
+\subsection{CPU and OS}
+The first thing specified in the OIL file is the CPU, which serves as a container for all other objects\cite{irisa25}.
+After the CPU, an OS is specified which is the object used to defined the setup of the operating system of the OSEK application.
+A CPU has to have exactly one OS object defined.
+The OS has a series of attributes: 
+\subsubsection*{Status}
+Status specifies the level of error checking, which can be set to either \textit{standard} or \textit{extended}, where extended allows for additional return values for tasks, such as \textit{task is invalid} or \textit{task still occupies resources} \cite{irisa223}.\\
+
+\subsubsection*{Hook routines}
+The OS has a series of boolean hook routines, meaning that they can either be enabled or disabled. 
+\textbf{STARTUPHOOKS} is the first routine to be called with the OS starts, where the user can place initialization procedures, during this hook routine all user interrupts are disabled \cite{irisa223}.
+The \textbf{ERRORHOOK} routine is called in case of application errors, with the purpose of treating status information in a centralized manner. 
+The \textbf{SHUTDOWNHOOK} routine is called by the operating system when the OS service \textit{ShutdownOS} has been called, meaning during the shutdown of the system. 
+The two task routines, \textbf{PRETASKHOOK} and \textbf{POSTTASKHOOK} are called on task context switches, meaning that they can be used for debugging or time measurement, including context switch time \cite{irisa223}.
```

Comment:
Generally, maybe for the task context switches and other "special" terms, maybe emphasize them a bit, like the other ones. Dunno, they stick out to me a bit. (task context switches and context switch time)

Reply:
I'm thinking that it was covered last semester, so it should be prior knowledge